File: src/utils/gameLogic/blocks.ts
import { BLOCK_PATTERNS, BOARD_WIDTH } from '@/constants/gameConfig';
import type {
  Block,
  BlockPattern,
  GameBoard,
  Position,
  FallingColumn,
} from '@/types/game';
import type { SeededRNGType } from '@/utils/seededRNG';

/**
 * Generate random block using seeded RNG
 */
export function generateRandomBlock(rng: SeededRNGType): Block {
  const pattern = rng.choice(BLOCK_PATTERNS) as BlockPattern;

  return {
    pattern,
    id: rng.generateId(),
  };
}

/**
 * Rotate a 2x2 block pattern clockwise
 */
export function rotateBlockPattern(
  pattern: BlockPattern,
  clockwise: boolean = true
): BlockPattern {
  const size = pattern.length;
  const rotated: BlockPattern = Array(size)
    .fill(null)
    .map(() => Array(size).fill(0));

  for (let i = 0; i < size; i++) {
    for (let j = 0; j < size; j++) {
      if (clockwise) {
        rotated[j][size - 1 - i] = pattern[i][j];
      } else {
        rotated[size - 1 - j][i] = pattern[i][j];
      }
    }
  }

  return rotated;
}

/**
 * Calculate available spaces for block placement in a column
 */
function getAvailableSpaces(
  boardX: number,
  fallingColumns: FallingColumn[],
  board: GameBoard
): number {
  const BLOCK_HEIGHT = 2;

  // Check space blocked by falling columns
  let spacesByFalling = BLOCK_HEIGHT;
  const existingColumn = fallingColumns.find(col => col.x === boardX);
  if (existingColumn && existingColumn.cells.length > 0) {
    const topCell = existingColumn.cells[existingColumn.cells.length - 1];
    spacesByFalling = Math.min(topCell.y, BLOCK_HEIGHT);
  }

  // Check space blocked by board cells
  let spacesByBoard = 0;
  for (let y = 0; y < BLOCK_HEIGHT; y++) {
    if (board[y][boardX] === 0) {
      spacesByBoard++;
    }
  }

  return Math.min(spacesByFalling, spacesByBoard);
}

/**
 * Place block on board with partial placement support
 * Only places cells that fit within bounds and don't collide
 */
export function placeBlockOnBoard(
  board: GameBoard,
  fallingColumns: FallingColumn[],
  block: Block,
  position: Position
): GameBoard {
  const newBoard = board.map(row => [...row]);
  const pattern = block.pattern;

  if (position.y < 0) {
    const BLOCK_HEIGHT = 2;

    for (let x = 0; x < BLOCK_HEIGHT; x++) {
      const boardX = position.x + x;
      const availableSpaces = getAvailableSpaces(boardX, fallingColumns, board);

      // Place cells from bottom up, only as many as fit
      for (let y = 1; y >= BLOCK_HEIGHT - availableSpaces; y--) {
        newBoard[y - (BLOCK_HEIGHT - availableSpaces)][boardX] = pattern[y][x];
      }
    }
  } else {
    for (let y = 0; y < pattern.length; y++) {
      for (let x = 0; x < pattern[y].length; x++) {
        if (pattern[y][x] === 0) {
          continue;
        }

        const boardX = position.x + x;
        if (boardX >= 0 && boardX < BOARD_WIDTH && newBoard[y][boardX] === 0) {
          newBoard[y][boardX] = pattern[y][x];
        }
      }
    }
  }

  return newBoard;
}

/**
 * Check if any part of a block can be placed on the board
 * Returns true if at least one cell of the block can be placed
 */
export function canPlaceAnyPartOfBlock(
  board: GameBoard,
  position: Position
): boolean {
  const firstRow = board[0];
  if (firstRow[position.x] === 0 || firstRow[position.x + 1] === 0) {
    return true;
  }
  return false;
}

File: python/game/blocks.py
"""
blocks.py — Block generation and placement.
Port of src/utils/gameLogic/blocks.ts
"""

import copy
from typing import List, Optional
from .constants import BLOCK_PATTERNS, BOARD_WIDTH, BLOCK_HEIGHT
from .types import Block, FallingColumn
from .rng import SeededRNG


def generate_random_block(rng: SeededRNG) -> Block:
    """Generate a random block using seeded RNG."""
    pattern = rng.choice(BLOCK_PATTERNS)
    block_id = rng.generate_id()
    return Block(
        pattern=[row[:] for row in pattern],
        id=block_id,
    )


def rotate_block_pattern(pattern: List[List[int]], clockwise: bool = True) -> List[List[int]]:
    """
    Rotate a 2×2 block pattern.
    Clockwise:     rotated[j][size-1-i] = pattern[i][j]
    Counter-CW:    rotated[size-1-j][i] = pattern[i][j]
    """
    size = len(pattern)
    rotated = [[0] * size for _ in range(size)]

    for i in range(size):
        for j in range(size):
            if clockwise:
                rotated[j][size - 1 - i] = pattern[i][j]
            else:
                rotated[size - 1 - j][i] = pattern[i][j]

    return rotated


def _get_available_spaces(
    board_x: int,
    falling_columns: List[FallingColumn],
    board: List[List[int]],
) -> int:
    """
    Calculate how many rows from the top of the board are available
    for a block column to be placed into (respecting falling cells and
    existing board content in the top BLOCK_HEIGHT rows).
    """
    # Space limited by falling columns at this x
    spaces_by_falling = BLOCK_HEIGHT
    for col in falling_columns:
        if col.x == board_x and col.cells:
            top_cell = col.cells[-1]  # cells sorted bottom-to-top; last = highest
            spaces_by_falling = min(top_cell.y, BLOCK_HEIGHT)
            break

    # Space limited by board cells in the top rows
    spaces_by_board = 0
    for y in range(BLOCK_HEIGHT):
        if board[y][board_x] == 0:
            spaces_by_board += 1

    return min(spaces_by_falling, spaces_by_board)


def place_block_on_board(
    board: List[List[int]],
    falling_columns: List[FallingColumn],
    block: Block,
    position_x: int,
    position_y: int,
) -> List[List[int]]:
    """
    Place a block on the board.
    If position_y < 0 (block is above the board), uses partial placement
    logic that respects falling columns as obstacles.
    Otherwise places cells that fit within bounds and don't collide.
    """
    new_board = [row[:] for row in board]

    if position_y < 0:
        return _place_block_above_board(new_board, block, position_x, position_y, falling_columns, board)
    else:
        return _place_block_normally(new_board, block, position_x, position_y)


def _place_block_above_board(
    new_board: List[List[int]],
    block: Block,
    position_x: int,
    position_y: int,
    falling_columns: List[FallingColumn],
    original_board: List[List[int]],
) -> List[List[int]]:
    pattern = block.pattern

    for x in range(BLOCK_HEIGHT):
        board_x = position_x + x
        available_spaces = _get_available_spaces(board_x, falling_columns, original_board)

        # Place cells from bottom up, only as many as fit
        for y in range(1, BLOCK_HEIGHT - available_spaces - 1, -1):
            if 0 <= y < BLOCK_HEIGHT and 0 <= board_x < BOARD_WIDTH:
                dest_y = y - (BLOCK_HEIGHT - available_spaces)
                if 0 <= dest_y < len(new_board):
                    new_board[dest_y][board_x] = pattern[y][x]

    return new_board


def _place_block_normally(
    new_board: List[List[int]],
    block: Block,
    position_x: int,
    position_y: int,
) -> List[List[int]]:
    pattern = block.pattern

    for y in range(len(pattern)):
        for x in range(len(pattern[y])):
            if pattern[y][x] == 0:
                continue
            board_x = position_x + x
            board_y = position_y + y
            if (0 <= board_x < BOARD_WIDTH and
                    0 <= board_y < len(new_board) and
                    new_board[board_y][board_x] == 0):
                new_board[board_y][board_x] = pattern[y][x]

    return new_board


def can_place_any_part_of_block(
    board: List[List[int]],
    position_x: int,
) -> bool:
    """
    Returns True if at least one cell of the block can be placed.
    Checks whether the first row has empty space at the block's x or x+1.
    """
    first_row = board[0]
    return first_row[position_x] == 0 or first_row[position_x + 1] == 0
